### PR TITLE
Fix #713: Define JSON deserialization

### LINKED ISSUE #713: JSON deserialization is not defined
The prose would read much clearer if this was an operation that returned something.

### PR DESCRIPTION
**Pull request by emlun**
This replaces the "perform JSON deserialization" language with references to _parse JSON from bytes_ in Infra. See #713.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/pull/718.html" title="Last updated on Jan 17, 2018, 11:34 AM GMT (910c25d)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/718/a6c0da2...910c25d.html" title="Last updated on Jan 17, 2018, 11:34 AM GMT (910c25d)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -2719,35 +2718,36 @@ When verifying a given {{PublicKeyCredential}} structure (|credential|) as part
 1. Using |credential|'s {{Credential/id}} attribute (or the corresponding {{PublicKeyCredential/rawId}}, if
     [=base64url encoding=] is inappropriate for your use case), look up the corresponding credential public key.
 
-2. Let |cData|, |aData| and |sig| denote the value of |credential|'s {{PublicKeyCredential/response}}'s
+1. Let |cData|, |aData| and |sig| denote the value of |credential|'s {{PublicKeyCredential/response}}'s
     {{AuthenticatorResponse/clientDataJSON}}, {{AuthenticatorAssertionResponse/authenticatorData}}, and
     {{AuthenticatorAssertionResponse/signature}} respectively.
 
-3. Perform JSON deserialization on |cData| to extract the [=client data=] |C| used for the signature.
+1. Let |C|, the [=client data=] used for the signature, be the result ofj[=parse JSON from
```
Comment (jcjones) [inline index.bs]: spurious j; `s/ofj/of/`

Comment (emlun) [reply index.bs]: Oops. Thanks!

**Comment by emlun** [comment]
@jcjones 
>I feel like we should do a style run and catch all these ordered list changes at once... and write a style guide.

Sounds like a good idea, we could rebrand this PR for that goal instead. Should we assign a milestone?

File: index.bs [hunk 2]
```diff
@@ -2619,67 +2619,66 @@ structures.
...
 [=registration=] [=ceremony=], a [=[RP]=] MUST proceed as follows:
 
-1. Perform JSON deserialization on the {{AuthenticatorResponse/clientDataJSON}} field of the
-    {{AuthenticatorAttestationResponse}} object to extract the [=client data=] |C| claimed as collected during the credential
-    creation.
+1. Let |C|, the [=client data=] claimed as collected during the credential creation, be the result of [=parse JSON from
+    bytes|parsing JSON from the bytes=] in <code>|response|.{{AuthenticatorResponse/clientDataJSON}}</code>.
 
-2. Verify that the {{CollectedClientData/type}} in |C| is the string `webauthn.create`.
+1. Verify that the {{CollectedClientData/type}} in |C| is the string `webauthn.create`.
```
Comment (equalsJeffH) [inline index.bs]: should this be:
> Verify that the value of \<code>|C|.{{CollectedClientData/type}}\</code> is \`webauthn.create\`.

..?

File: index.bs [hunk 3]
```diff
@@ -2619,67 +2619,66 @@ structures.
...
 
-2. Verify that the {{CollectedClientData/challenge}} in |C| matches the challenge that was sent to the authenticator in the
+1. Verify that the {{CollectedClientData/challenge}} in |C| matches the challenge that was sent to the authenticator in the
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

File: index.bs [hunk 4]
```diff
@@ -2619,67 +2619,66 @@ structures.
...
     {{CredentialsContainer/create()}} call.
 
-3. Verify that the {{CollectedClientData/origin}} in |C| matches the [=[RP]=]'s [=origin=].
+1. Verify that the {{CollectedClientData/origin}} in |C| matches the [=[RP]=]'s [=origin=].
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

File: index.bs [hunk 5]
```diff
@@ -2619,67 +2619,66 @@ structures.
...
 
-4. Verify that the {{CollectedClientData/tokenBindingId}} in |C| matches the [=Token Binding ID=] for the TLS connection over
+1. Verify that the {{CollectedClientData/tokenBindingId}} in |C| matches the [=Token Binding ID=] for the TLS connection over
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

File: index.bs [hunk 6]
```diff
@@ -2619,67 +2619,66 @@ structures.
...
     which the attestation was obtained.
 
-5. Verify that the {{CollectedClientData/clientExtensions}} in |C| is a subset of the extensions requested by the RP
+1. Verify that the {{CollectedClientData/clientExtensions}} in |C| is a subset of the extensions requested by the RP
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

File: index.bs [hunk 7]
```diff
@@ -2619,67 +2619,66 @@ structures.
...
-6. Compute the hash of {{AuthenticatorResponse/clientDataJSON}} using SHA-256.
+1. Compute the hash of <code>|response|.{{AuthenticatorResponse/clientDataJSON}}</code> using SHA-256.
 
-7. Perform CBOR decoding on the {{AuthenticatorAttestationResponse/attestationObject}} field of the
+1. Perform CBOR decoding on the {{AuthenticatorAttestationResponse/attestationObject}} field of the
     {{AuthenticatorAttestationResponse}} structure to obtain the attestation statement format |fmt|, the [=authenticator data=]
     |authData|, and the attestation statement |attStmt|.
 
-8. Verify that the [=RP ID=] hash in |authData| is indeed the SHA-256 hash of the [=RP ID=] expected by the RP.
+1. Verify that the [=RP ID=] hash in |authData| is indeed the SHA-256 hash of the [=RP ID=] expected by the RP.
```
Comment (equalsJeffH) [inline index.bs]: suggest:
> Verify that \<code>|authData|.rpIdHash\</code> is indeed the SHA-256 hash of the [=RP ID=] expected by the RP.

..?

File: index.bs [hunk 8]
```diff
@@ -2705,34 +2704,35 @@ When verifying a given {{PublicKeyCredential}} structure (|credential|) as part
...
 
-3. Perform JSON deserialization on |cData| to extract the [=client data=] |C| used for the signature.
+1. Let |C|, the [=client data=] used for the signature, be the result of [=parse JSON from
+    bytes|parsing JSON from the bytes=] in |cData|.
 
-3. Verify that the {{CollectedClientData/type}} in |C| is the string `webauthn.get`.
+1. Verify that the {{CollectedClientData/type}} in |C| is the string `webauthn.get`.
 
-4. Verify that the {{CollectedClientData/challenge}} member of |C| matches the challenge that was sent to the authenticator in
+1. Verify that the {{CollectedClientData/challenge}} member of |C| matches the challenge that was sent to the authenticator in
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

File: index.bs [hunk 9]
```diff
@@ -2705,34 +2704,35 @@ When verifying a given {{PublicKeyCredential}} structure (|credential|) as part
...
     the {{PublicKeyCredentialRequestOptions}} passed to the {{CredentialsContainer/get()}} call.
 
-5. Verify that the {{CollectedClientData/origin}} member of |C| matches the [=[RP]=]'s [=origin=].
+1. Verify that the {{CollectedClientData/origin}} member of |C| matches the [=[RP]=]'s [=origin=].
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

File: index.bs [hunk 10]
```diff
@@ -2705,34 +2704,35 @@ When verifying a given {{PublicKeyCredential}} structure (|credential|) as part
...
 
-6. Verify that the {{CollectedClientData/tokenBindingId}} member of |C| (if present) matches the [=Token Binding ID=] for the
+1. Verify that the {{CollectedClientData/tokenBindingId}} member of |C| (if present) matches the [=Token Binding ID=] for the
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

File: index.bs [hunk 11]
```diff
@@ -2705,34 +2704,35 @@ When verifying a given {{PublicKeyCredential}} structure (|credential|) as part
...
     TLS connection over which the signature was obtained.
 
-7. Verify that the {{CollectedClientData/clientExtensions}} member of |C| is a subset of the extensions requested by the
+1. Verify that the {{CollectedClientData/clientExtensions}} member of |C| is a subset of the extensions requested by the
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

File: index.bs [hunk 12]
```diff
@@ -2705,34 +2704,35 @@ When verifying a given {{PublicKeyCredential}} structure (|credential|) as part
...
     [=[RP]=] and that the {{CollectedClientData/authenticatorExtensions}} in |C| is also a subset of the extensions
     requested by the [=[RP]=].
 
-8. Verify that the <code>[=rpIdHash=]</code> in |aData| is the SHA-256 hash of the [=RP ID=] expected by the [=[RP]=].
+1. Verify that the <code>[=rpIdHash=]</code> in |aData| is the SHA-256 hash of the [=RP ID=] expected by the [=[RP]=].
```
Comment (equalsJeffH) [inline index.bs]: ditto here?

**Comment by equalsJeffH** [comment]
reviewed https://github.com/w3c/webauthn/pull/718#pullrequestreview-83906027

File: index.bs (hunk 7 above)
Comment (emlun) [reply index.bs]: My thinking in writing this one on the "the X in |authData|" format was to hint at the fact that the representation of |authData| doesn't separate the semantic contents into neat members like JSON or CBOR does. What do you think of that? I'm not against changing it if you still think it's better to write it on the "|authData|.X" format.

File: index.bs (hunk 12 above)
Comment (emlun) [reply index.bs]: See reply above.

**Comment by emlun** [comment]
@equalsJeffH Thanks for the review, good points! I replied to two of them, see https://github.com/w3c/webauthn/pull/718#discussion_r158018877 .

**Comment by emlun** [comment]
Actually, come to think of it, this is the RP operations. Perhaps it's confusing to refer to procedures specific to web browsers in this section that should be environment agnostic?

**Comment by emlun** [comment]
@equalsJeffH What do you think about https://github.com/w3c/webauthn/pull/718#issuecomment-355076830 ?

File: index.bs [hunk 13]
```diff
@@ -2637,28 +2637,27 @@ structures.
...
 
-1. Verify that the {{CollectedClientData/clientExtensions}} in |C| is a subset of the extensions requested by the RP
-    and that the {{CollectedClientData/authenticatorExtensions}} in |C| is also a subset of the extensions requested by
-    the RP.
+1. Verify that the value of <code>|C|.{{CollectedClientData/clientExtensions}}</code> is a subset of the extensions requested by
+    the RP and that the value of <code>|C|.{{CollectedClientData/authenticatorExtensions}}</code> is also a subset of the
+    extensions requested by the RP.
 
-1. Compute the hash of {{AuthenticatorResponse/clientDataJSON}} using SHA-256.
+1. Compute the hash of <code>|resoinse|.{{AuthenticatorResponse/clientDataJSON}}</code> using SHA-256.
```
Comment (jcjones) [inline index.bs]: s/resoinse/response/

**Comment by nadalin** [comment]
@emlun Are we ready to merge ?

**Comment by emlun** [comment]
@nadalin Not quite. I'm not sure we should merge this (see https://github.com/w3c/webauthn/pull/718#issuecomment-355076830), and if we should I need to address @jcjones's review comment.

**Comment by emlun** [comment]
Closing this to isolate sets of unrelated changes in new PRs.

**Pull request closed without merging**